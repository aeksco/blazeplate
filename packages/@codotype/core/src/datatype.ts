export enum Datatype {
    // Standard datatypes
    STRING = "STRING", // Postgres, MongoDB
    TEXT = "TEXT", // Postgres, MongoDB (Equiv. to String)
    INT = "INT", // Postgres, MongoDB
    BIGINT = "BIGINT", // Postgres
    FLOAT = "FLOAT", // Postgres
    DECIMAL = "DECIMAL", // Postgres
    NUMERIC = "NUMERIC", // Postgres
    BOOLEAN = "BOOLEAN", // Postgres, MongoDB
    DATE = "DATE", // Postgres, MongoDB
    TIME = "TIME", // Postgres
    DATETIME = "DATETIME", // Postgres
    TIMESTAMP = "TIMESTAMP", // MongoDB
    BINARY = "BINARY", // Postgres, MongoDB
    JSON = "JSON", // Postgres
    JSONB = "JSONB", // Postgres
    OBJECT = "OBJECT", // MongoDB
    ID = "ID", // Postgres, MongoDB
    UUID = "UUID", // Postgres
    AUTO_INCREMENTED_ID = "AUTO_INCREMENTED_ID", // Postgres
    OBJECT_ID = "OBJECT_ID", // MongoDB

    // Array datatypes
    STRING_ARRAY = "STRING_ARRAY", // Postgres, MongoDB
    TEXT_ARRAY = "TEXT_ARRAY", // Postgres, MongoDB (Equiv. to String)
    INT_ARRAY = "INT_ARRAY", // Postgres, MongoDB
    BIGINT_ARRAY = "BIGINT_ARRAY", // Postgres
    FLOAT_ARRAY = "FLOAT_ARRAY", // Postgres
    DECIMAL_ARRAY = "DECIMAL_ARRAY", // Postgres
    NUMERIC_ARRAY = "NUMERIC_ARRAY", // Postgres
    BOOLEAN_ARRAY = "BOOLEAN_ARRAY", // Postgres, MongoDB
    DATE_ARRAY = "DATE_ARRAY", // Postgres, MongoDB
    TIME_ARRAY = "TIME_ARRAY", // Postgres
    DATETIME_ARRAY = "DATETIME_ARRAY", // Postgres
    TIMESTAMP_ARRAY = "TIMESTAMP_ARRAY", // MongoDB
    UUID_ARRAY = "UUID_ARRAY", // Postgres
    OBJECT_ID_ARRAY = "OBJECT_ID_ARRAY", // MongoDB

    // Special Datatypes
    SINGLE_FILE = "SINGLE_FILE",
    SINGLE_IMAGE = "SINGLE_IMAGE",

    // Future datatypes?
    // BLOB = "BLOB";
    // BYTEA = "BYTEA";
    // CHAR = "CHAR";
    // CHAR = "CHAR";
    // DOUBLE = "DOUBLE";
    // ENUM = "ENUM";
    // GEOMETRY = "GEOMETRY";
    // GEOMETRY_COLLECTION = "GEOMETRY_COLLECTION";
    // LINESTRING = "LINESTRING";
    // LONGTEXT = "LONGTEXT";
    // MEDIUM_INT = "MEDIUM_INT";
    // MEDIUM_TEXT = "MEDIUM_TEXT";
    // MULTI_LINE_STRING = "MULTI_LINE_STRING";
    // MULTI_POINT = "MULTI_POINT";
    // MULTI_POLYGON = "MULTI_POLYGON";
    // POINT = "POINT";
    // POLYGON = "POLYGON";
    // SMALL_INT = "SMALL_INT";
    // TIMESTAMPTZ = "TIMESTAMPTZ";
    // TIMETZ = "TIMETZ";
    // TINY_INT = "TINY_INT";
    // VARCHAR = "VARCHAR";
    // YEAR = "YEAR";
}
