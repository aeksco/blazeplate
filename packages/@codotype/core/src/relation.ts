import { UUID } from "./uuid";
import { TokenPluralization } from "./token";
import { SchemaCreators } from "./schema";

// // // //

/**
 * RelationType
 * Defines an enum of available RelationsTypes
 */
export type RelationType =
    | "BELONGS_TO"
    | "HAS_ONE"
    | "HAS_MANY"
    | "TO_ONE"
    | "TO_MANY"
    | "HAS_AND_BELONGS_TO_MANY"
    | "EMBEDS_ONE"
    | "EMBEDS_MANY";
// TODO - update enum casing to:
// RelationTypes.belongsTo
// RelationTypes.hasOne
// RelationTypes.hasMany
// RelationTypes.hasAndBelongsToMany
// RelationTypes.toOne
// RelationTypes.toMany
// RelationTypes.embedsOne
// RelationTypes.embedsMany
export enum RelationTypes {
    BELONGS_TO = "BELONGS_TO",
    HAS_ONE = "HAS_ONE",
    HAS_MANY = "HAS_MANY",
    TO_ONE = "TO_ONE",
    TO_MANY = "TO_MANY",
    HAS_AND_BELONGS_TO_MANY = "HAS_AND_BELONGS_TO_MANY",
    EMBEDS_ONE = "EMBEDS_ONE",
    EMBEDS_MANY = "EMBEDS_MANY",
}

/**
 * RelationInput
 * Defines the RelationInput interface
 * TODO - add support for addons here -> see AttributeAddon for context
 */
export interface RelationInput {
    id: UUID;
    type: RelationType;
    required: boolean; // TODO - keep this for now - should this be handled as an Addon? YES.
    source: SchemaCreators;
    internalNote: string;
    sourceSchemaID: UUID;
    destinationSchemaID: UUID;
    sourceSchemaAlias: string; // TODO - make this TokenPluralization?
    destinationSchemaAlias: string; // TODO - make this TokenPluralization?
}

// Relation constants
export const DEFAULT_RELATION_INPUT: RelationInput = {
    id: "",
    type: null,
    required: false,
    source: SchemaCreators.user,
    internalNote: "",
    sourceSchemaID: "",
    destinationSchemaID: "",
    sourceSchemaAlias: "",
    destinationSchemaAlias: "",
};

// // // //

/**
 * Relation
 * Defines the Reference interface
 * TODO - add support for addons here -> see AttributeAddon for context
 */
export interface Relation {
    id: UUID;
    type: RelationType;
    sourceRelationId: UUID;
    sourceSchemaId: UUID;
    destinationSchemaId: UUID;
    identifiers: {
        source: {
            canonical: TokenPluralization;
            alias: TokenPluralization;
        };
        destination: {
            canonical: TokenPluralization;
            alias: TokenPluralization;
        };
    };
}
