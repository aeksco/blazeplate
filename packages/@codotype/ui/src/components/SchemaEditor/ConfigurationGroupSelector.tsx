import * as React from "react";
import { ConfigurationInput } from "../ConfigurationInput";
import {
    ConfigurationGroup,
    ConfigurationPropertyDict,
    ConfigurationValue,
} from "@codotype/core";

// // // //

export function ConfigurationGroupTab(props: {
    label: string;
    active: boolean;
    onClick: () => void;
}) {
    const { label } = props;
    const btnClassName: string[] = ["btn"];
    if (props.active) {
        btnClassName.push("btn-light active");
    } else {
        btnClassName.push("btn-light");
    }

    return (
        <div className="d-flex mr-2">
            <button
                className={btnClassName.join(" ")}
                onClick={(e) => {
                    e.currentTarget.blur();
                    props.onClick();
                }}
            >
                {label}
            </button>
        </div>
    );
}

/**
 * ConfigurationGroupSelector
 * @param props.project
 * @param props.PluginMetadata
 * @param props.onChange
 */
export function ConfigurationGroupSelector(props: {
    configuration: ConfigurationValue;
    configurationGroups: ConfigurationGroup[];
    children: React.ReactNode;
    onChange: (updatedPluginConfiguration: ConfigurationValue) => void;
}) {
    const { configurationGroups } = props;

    // Gets default ConfigurationGroup to render
    const defaultConfigurationGroup: ConfigurationGroup | undefined =
        configurationGroups[0];

    // If there is no default ConfigurationGroup -> just return props.children
    if (!defaultConfigurationGroup) {
        return <React.Fragment>{props.children}</React.Fragment>;
    }

    // Stores the currently selected ConfigurationGroup
    const [
        selectedConfigurationGroup,
        selectConfigurationGroup,
    ] = React.useState<ConfigurationGroup>(defaultConfigurationGroup);

    // NOTE - enable/disable this if schemas aren't supported
    const [viewingSchemas, setViewingSchemas] = React.useState<boolean>(true);

    return (
        <div className="row">
            <div className="col-lg-12">
                <div className="d-flex flex-row mt-3 mb-2">
                    <ConfigurationGroupTab
                        onClick={() => {
                            setViewingSchemas(true);
                        }}
                        active={viewingSchemas}
                        label={"Schema"}
                    />

                    {/* Renders the navigation for selecting a ConfigurationGroup */}
                    {configurationGroups.map(
                        (configurationGroup: ConfigurationGroup) => {
                            return (
                                <ConfigurationGroupTab
                                    key={configurationGroup.identifier}
                                    onClick={() => {
                                        setViewingSchemas(false);
                                        selectConfigurationGroup(
                                            configurationGroup,
                                        );
                                    }}
                                    active={
                                        configurationGroup.identifier ===
                                        selectedConfigurationGroup.identifier &&
                                        !viewingSchemas
                                    }
                                    label={configurationGroup.content.label}
                                />
                            );
                        },
                    )}
                </div>
            </div>
            <div className="col-lg-12">
                {/* Renders the ConfigurationInput */}
                {!viewingSchemas && (
                    <ConfigurationInput
                        configurationGroup={selectedConfigurationGroup}
                        value={
                            props.configuration[
                            selectedConfigurationGroup.identifier
                            ]
                        }
                        onChange={(updatedVal: ConfigurationPropertyDict) => {
                            // Defines updatd project with latest configuration value
                            const updatedPluginConfiguration: ConfigurationValue = {
                                ...props.configuration,
                                [selectedConfigurationGroup.identifier]: updatedVal,
                            };

                            // Invokes props.onChange with updated project
                            props.onChange(updatedPluginConfiguration);
                        }}
                    />
                )}

                {/* Render SchemaEditorLayout */}
                {viewingSchemas && <div className="mt-3">{props.children}</div>}
            </div>
            {/* <pre>{JSON.stringify(val, null, 4)}</pre> */}
            {/* <pre>{JSON.stringify(props.configurationGroup, null, 4)}</pre> */}
        </div>
    );
}
