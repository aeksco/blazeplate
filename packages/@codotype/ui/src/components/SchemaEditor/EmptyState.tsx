import * as React from "react";
import { SchemaForm } from "./SchemaForm";

// // // //

/**
 * EmptyState
 */
export function EmptyState(props: {
    disabled: boolean;
    onNewTokenChange: (updatedTokens: any) => void;
    onKeydownEnter: () => void;
    onSubmit: () => void;
}) {
    const { disabled, onNewTokenChange, onKeydownEnter, onSubmit } = props;
    return (
        <div className="flex py-4 justify-center items-center">
            <div className="grid grid-cols-5">
                <div className="col-start-2 col-span-3">
                    <div className="card card-body shadow-xl">
                        <div className="grid grid-cols-12">
                            <div className="col-span-12">
                                <h4 className="mb-0 text-2xl flex">
                                    New Data Model
                                </h4>

                                <SchemaForm
                                    label={""}
                                    onChange={updatedTokens => {
                                        onNewTokenChange(updatedTokens);
                                    }}
                                    onKeydownEnter={() => {
                                        onKeydownEnter();
                                    }}
                                />
                            </div>
                        </div>
                        <div className="col-span-12 text-right mt-2">
                            <button
                                className="btn btn-lg btn-primary"
                                disabled={disabled}
                                onClick={() => {
                                    onSubmit();
                                }}
                            >
                                Create Schema
                            </button>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    );
}
