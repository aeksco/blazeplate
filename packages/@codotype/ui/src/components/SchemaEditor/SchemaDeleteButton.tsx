import * as React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTrashAlt, faInfoCircle } from "@fortawesome/free-solid-svg-icons";
import { OverlayTrigger, Tooltip } from "react-bootstrap";
import styled from "styled-components";
import classnames from "classnames";
import {
    Schema,
    CreatedByValues,
    inflateSchema,
    SchemaInput,
    ProjectInput,
    buildRelations,
} from "@codotype/core";

// // // //

const StyledButton = styled.button`
    transition: color 0.25s ease-in-out;
    cursor: pointer;
    color: #adb5bd;
    box-shadow: none;
    color: #d9534f;
    &.hasReverse {
        &:hover {
            color: #adb5bd !important;
            cursor: not-allowed;
        }
    }
    &:hover {
        color: #d9534f;
    }

    &.disable-submit {
        cursor: not-allowed !important;
        &:hover {
            color: #adb5bd;
        }
    }
`;

const StyledButtonDisabled = styled.button``;

/**
 * SchemaDeleteButton
 * @param props.onClick
 */
export function SchemaDeleteButton(props: {
    schemaInput: SchemaInput;
    projectInput: ProjectInput;
    onClick: () => void;
}) {
    const { schemaInput, projectInput } = props;

    const inflatedSchema: Schema = inflateSchema({
        schemaInput,
        relations: buildRelations({
            schemaInputs: projectInput.schemas,
            relationInputs: projectInput.relations,
        }),
    });

    // Defines boolean indicating whether or not the schema can be removed
    const disableSubmit: boolean =
        inflatedSchema.referencedBy.filter(
            r => r.sourceSchemaID !== schemaInput.id,
        ).length > 0;

    if (
        schemaInput.createdBy === CreatedByValues.plugin &&
        schemaInput.locked
    ) {
        return (
            <OverlayTrigger
                placement="left"
                overlay={
                    <Tooltip id="auto-generated-schema-tooltip">
                        The{" "}
                        <strong>
                            {schemaInput.identifiers.singular.title}
                        </strong>{" "}
                        Schema is auto-generated and may not be edited.
                    </Tooltip>
                }
            >
                <span
                    className="badge badge-success ml-2"
                    style={{ cursor: "default" }}
                >
                    <FontAwesomeIcon className="mr-2" icon={faInfoCircle} />
                    Auto-Generated
                </span>
            </OverlayTrigger>
        );
    }

    let tooltipContent = (
        <React.Fragment>
            Remove the <strong>{schemaInput.identifiers.singular.title}</strong>{" "}
            Schema.
        </React.Fragment>
    );

    if (disableSubmit) {
        tooltipContent = (
            <React.Fragment>
                You cannot remove the{" "}
                <strong>{schemaInput.identifiers.singular.title}</strong> Schema
                because it is referenced by a relation on another Schema.
            </React.Fragment>
        );
    }

    // Render different layout + tooltip if disableSubmit
    return (
        <OverlayTrigger
            placement="left"
            overlay={
                <Tooltip id={`delete-button-tooltip-${schemaInput.id}`}>
                    {tooltipContent}
                </Tooltip>
            }
        >
            <StyledButton
                className={classnames("btn btn-link ml-2 py-0", {
                    "disable-submit": disableSubmit,
                })}
                onClick={() => {
                    if (disableSubmit) {
                        return;
                    }
                    props.onClick();
                }}
            >
                <FontAwesomeIcon className="ml-1" icon={faTrashAlt} />
            </StyledButton>
        </OverlayTrigger>
    );
}
