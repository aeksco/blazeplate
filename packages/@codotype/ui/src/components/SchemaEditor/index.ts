export { SchemaEditorLayout } from "./component";
export { SchemaDeleteModal } from "./SchemaDeleteModal";
export { SchemaDeleteButton } from "./SchemaDeleteButton";
export { SchemaDetail } from "./SchemaDetail";
export { SchemaDetailHeader } from "./SchemaDetailHeader";
export { SchemaEditButton } from "./SchemaEditButton";
export { SchemaForm } from "./SchemaForm";
export { SchemaFormModal } from "./SchemaFormModal";
export { SchemaIncomingRelations } from "./SchemaIncomingRelations";
export { SchemaNewButton } from "./SchemaNewButton";
export { SchemaPreview } from "./SchemaPreview";
export { ConfigurationGroupSelector } from "./ConfigurationGroupSelector";
export { SchemaSelector } from "./SchemaSelector";
