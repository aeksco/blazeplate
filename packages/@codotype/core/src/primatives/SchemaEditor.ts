import { SchemaEditorConfiguration } from "..";
import { Datatype } from "../datatype";
import { RelationTypes, RelationInput } from "../relation";
import { AttributeAddon, RelationAddon } from "../schema-editor-addon";
import { SchemaInput } from "../schema";
import { Attribute } from "../attribute";
import { ConfigurationGroup } from "../configuration-option-types";

// // // //

interface SchemaEditorBuilderParams {
    supportedDatatypes: Datatype[];
    supportedRelations: RelationTypes[];
    enableAttributeDefaultValue?: boolean;
    documentation?: string;
    defaultSchemas?: SchemaInput[];
    attributeAddons?: AttributeAddon[];
    defaultAttributes?: Attribute[];
    defaultRelations?: RelationInput[];
    configurationGroups?: ConfigurationGroup[];
}

export class SchemaEditorBuilder implements SchemaEditorConfiguration {
    supportedDatatypes: Datatype[];
    supportedRelations: RelationTypes[];

    enableAttributeDefaultValue: boolean = false;
    documentation: string = "";
    defaultSchemas: SchemaInput[] = [];
    attributeAddons: AttributeAddon[] = [];
    defaultAttributes: Attribute[] = [];
    defaultRelations: RelationInput[] = [];
    configurationGroups: ConfigurationGroup[] = [];

    constructor(params: SchemaEditorBuilderParams) {
        this.supportedDatatypes = params.supportedDatatypes;
        this.supportedRelations = params.supportedRelations;

        this.enableAttributeDefaultValue =
            params.enableAttributeDefaultValue !== undefined
                ? params.enableAttributeDefaultValue
                : this.enableAttributeDefaultValue;
        this.documentation = params.documentation || this.documentation;
        this.defaultSchemas = params.defaultSchemas || this.defaultSchemas;
        this.attributeAddons = params.attributeAddons || this.attributeAddons;
        this.defaultAttributes =
            params.defaultAttributes || this.defaultAttributes;
        this.defaultRelations =
            params.defaultRelations || this.defaultRelations;
        this.configurationGroups =
            params.configurationGroups || this.configurationGroups;
    }
}
