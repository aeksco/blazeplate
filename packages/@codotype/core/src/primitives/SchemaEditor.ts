import { AttributeInput, SchemaEditorConfiguration } from "..";
import { Datatype } from "../datatype";
import { RelationTypes, RelationInput } from "../relation";
import { AttributeAddon, RelationAddon } from "../schema-editor-addon";
import { SchemaInput } from "../schema";
import { ConfigurationGroup } from "../configuration-option-types";

// // // //

interface SchemaEditorBuilderParams {
    supportedDatatypes: Datatype[];
    supportedRelations: RelationTypes[];
    defaultSchemas?: SchemaInput[];
    attributeAddons?: AttributeAddon[];
    relationAddons?: RelationAddon[];
    defaultRelations?: RelationInput[];
    configurationGroups?: ConfigurationGroup[];
    newSchemaDefaints?: ConfigurationGroup[];
    newSchemaInputDefaults?: {
        attributes: AttributeInput[];
        relations: RelationInput[];
    };
}

const defaultSchemaInputDefaults = {
    attributes: [],
    relations: [],
};

export class SchemaEditorBuilder implements SchemaEditorConfiguration {
    supportedDatatypes: Datatype[];
    supportedRelationTypes: RelationTypes[];
    defaultSchemas: SchemaInput[] = [];
    attributeAddons: AttributeAddon[] = [];
    relationAddons: RelationAddon[] = [];
    defaultRelations: RelationInput[] = [];
    configurationGroups: ConfigurationGroup[] = [];
    newSchemaDefaults: {
        attributes: AttributeInput[];
        relations: RelationInput[];
    } = {
        ...defaultSchemaInputDefaults,
    };

    constructor(params: SchemaEditorBuilderParams) {
        this.supportedDatatypes = params.supportedDatatypes;
        this.supportedRelationTypes = params.supportedRelations;

        this.defaultSchemas = params.defaultSchemas || this.defaultSchemas;
        this.attributeAddons = params.attributeAddons || this.attributeAddons;
        this.relationAddons = params.relationAddons || this.relationAddons;
        this.defaultRelations =
            params.defaultRelations || this.defaultRelations;
        this.configurationGroups =
            params.configurationGroups || this.configurationGroups;

        if (params.newSchemaInputDefaults !== undefined) {
            this.newSchemaDefaults = {
                attributes: this.newSchemaDefaults.attributes,
                relations: this.newSchemaDefaults.relations,
            };
        }
    }
}
