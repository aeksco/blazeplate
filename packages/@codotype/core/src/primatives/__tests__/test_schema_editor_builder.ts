import { SchemaEditorBuilder } from "../SchemaEditor";
import { Datatype } from "../../datatype";
import { RelationTypes } from "../../relation";

// // // //

describe("factory", () => {
    test("works", () => {
        const schemaEditor = new SchemaEditorBuilder({
            supportedDatatypes: [Datatype.STRING],
            supportedRelations: [RelationTypes.TO_ONE],
        });

        expect(schemaEditor).toMatchSnapshot();
    });
});
