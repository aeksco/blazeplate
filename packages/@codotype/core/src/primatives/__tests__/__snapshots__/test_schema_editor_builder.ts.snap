// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factory works 1`] = `
SchemaEditorBuilder {
  "attributeAddons": Array [],
  "configurationGroups": Array [],
  "defaultAttributes": Array [],
  "defaultRelations": Array [],
  "defaultSchemas": Array [],
  "documentation": "",
  "enableAttributeDefaultValue": false,
  "supportedDatatypes": Array [
    "STRING",
  ],
  "supportedRelations": Array [
    "TO_ONE",
  ],
}
`;
